rfd/0182-multi-port-tcp-app-access.md

```diff
@@ -0,0 +1,544 @@
+the port number in the first few bytes of the connection and then proxy the rest of the downstream
+connection (of whatever client that wants to connect to a TCP app over a local proxy).
+
+A server speaking the custom protocol on the other end of the connection, an app agent, would read
+the port number and proxy the rest of the connection to the app itself on the given port.
+
+This all seems fine until we consider how we could support both multi-port apps and regular
+single-port apps. The ALPN proxy, after recognizing that the client wants to use
+`teleport-tcp-multi-port`, would need to forward the connection to an app agent through the reverse
+tunnel. For both types of TCP apps, the code that handles the connection would be pretty much the
+same, with the exception of reading the port first. Unless we changed the setup so that two kinds of
+app services register themselves in the reverse tunnel, the ALPN proxy would need to pass the
+information about the multi-port protocol out of band to the app service. This seems to defeat the
+purpose of using a custom protocol. The SNI and ALPN protocol solutions already pass the port number
+out of band, without the overhead of a custom protocol.
+
+##### Multiple apps with the same URI but different ports
+
+Instead of including port numbers in the definition of a single app, the admin could add multiple
+apps to the same app agent where the main difference between the apps would be a different port in
+the URI. Then VNet could somehow wrap those apps so that from the perspective of the user, they
+wouldn't need to connect to something like `app-1337.teleport.example.com` but rather just
+`app.teleport.example.com:1337`.
+
+While this could technically also work for web apps in the future, implementing this means adjusting
+every single user-facing tool to support multi-port apps (the Web UI, Connect, tsh, tctl, …). Old
+clients that do not support multi-port would see an almost identical app repeated once for every
+port. Supporting wide port ranges would also be significantly harder, as each port would require its
+own app resource. Per-session MFA would require creating a separate cert for each port.
+
+### Passing the port number from VNet to the client
+
+VNet [creates a proxy for an app](https://github.com/gravitational/teleport/blob/0b3163e7bd70afa5a2ddb5bd42f3efc034377d70/lib/vnet/app_resolver.go#L355)
+at the point of [resolving an A query for the app](https://github.com/gravitational/teleport/blob/0b3163e7bd70afa5a2ddb5bd42f3efc034377d70/lib/vnet/vnet.go#L567).
+With multi-port, VNet needs to create a local proxy at a later step, [in `vnet.NetworkStack.handleTCP`](https://github.com/gravitational/teleport/blob/b5df678e22d26776aa87f9dd493b3c443e013fa4/lib/vnet/vnet.go#L377),
+where it can read the port used by the connection (`req.ID().LocalPort`). It needs to maintain a map
+of port numbers to local proxies to avoid creating a new proxy and a cert on each connection.
```

Comment:
might be worth a mention of the intended strategy for VNet to select an app:port and how we expect that to behave and perform

VNet would still have to query for the app spec by the public addr while resolving the DNS A (or AAAA) request. At this point we would not know the port that the client will later want to dial to. Should we support multiple apps with the same public_addr and different port ranges? I think it's possible

We'll probably hold on to this app spec in the tcpAppHandler, and when receiving an incoming TCP connection, check if the port matches the app spec. A question is will we ever update this app spec if ports get added to it or removed? Could we try to update it if we get a request on a port we don't know about? this could be an improvement after the initial implementation.